File: dragonfly/noun.py
import typing
import copy

import dragonfly

from PyQt5.QtXml import QDomElement, QDomNode

class Noun(dragonfly.Entity):
	"""Nouns represents the objects of the game. Can be contained by other nouns.
		and has attributes and variables.
	"""

	id_max = 0

	def __init__(self, container: "Noun" = None) -> None:
		super().__init__()
		self.__container = container
		self.__attrs = set()
		self.__variables = dict()

		self.__before = []
		self.__after = []

		self.__connections = []

		# Generate an id
		Noun.id_max += 1
		self.__id = Noun.id_max

	def __str__(self) -> str:
		result = super().__str__()
		if self.__container != None:
			result += f" ({self.__container.name})"

		if self.__attrs:
			result += f" attrs: {', '.join(self.__attrs)}"
		
		return result

	@property
	def id(self) -> int:
		return self.__id

	@property
	def container(self) -> "Noun":
		"""Get the container Noun"""
		return self.__container

	@container.setter
	def container(self, cont: "Noun") -> None:
		"""Set the container noun. If the container is self, raise an exception."""
		if self == cont:
			raise Exception(f"Noun {self.name} cannot contain itself.")
		self.__container = cont

	def contains(self, name: str) -> bool:
		"""Check if this noun contains a noun.

		Args:
			name (str): the name of the child.

		Returns:
			bool: True if and only if this contains the child.
		"""
		for n in self.dictionary.nouns():
			if n.container == self:
				if n.responds(name): return True

		return False

	def childs(self, name: str = "") -> typing.List["Noun"]:
		"""Return a list with chils wich responds to name.
		If the name = "", return all childs.
		"""
		result = []
		for n in self.dictionary.nouns():
			if n.container == self:
				if not name:
					result.append(n)
				else:
					if n.responds(name):
						result.append(n)
		
		return result

	@property
	def attributes(self) -> typing.Set:
		"""Return a set with all attributes of the noun.
		"""
		return self.__attrs

	def set(self, values: typing.List[str]) -> None:
		"""Set a list of attributes."""
		self.__attrs.update(tuple(values))

	def isSet(self, value: str) -> bool:
		"""Check if an attribute is setted."""
		return value in self.__attrs

	def unset(self, values: typing.List[str]) -> None:
		"""Unset a list of attributes."""
		for v in values:
			self.__attrs.discard(v)

	@property
	def variables(self) -> typing.Dict:
		"""Return a dictionary with all variables of the noun.
		"""
		return self.__variables

	def getVariable(self, name: str) -> str:
		"""Return the value of the variable.
		Args:
			name (str): name of the variable.

		Returns:
			str: the value of the variable.
		"""
		if name in self.__variables.keys():
			return self.__variables[name]

		return ""

	def setVariable(self, name: str, value: str) -> None:
		"""Set the value of the variable.

		Args:
			name (str): name of the variable.
			value (str): value of the variable.
		"""
		self.__variables[name] = value

	@property
	def beforeEvents(self) -> typing.List["dragonfly.ActionEvent"]:
		"""Return the list of the Before events.

		Returns:
			dragonfly.ActionEvent: Before ActionEvent list.
		"""
		return self.__before
	
	@property
	def afterEvents(self) -> typing.List["dragonfly.ActionEvent"]:
		"""Return the list of the After events.

		Returns:
			dragonfly.ActionEvent: After ActionEvent list.
		"""
		return self.__after

	def addBefore(self, actionEvent: "dragonfly.ActionEvent") -> None:
		"""Add a new Before action evento to the list.

		Args:
			actionEvent (dragonfly.ActionEvent): Before ActionEvent.
		"""
		self.__before.append(actionEvent)

	def addAfter(self, actionEvent: "dragonfly.ActionEvent") -> None:
		"""Add a new After action evento to the list.

		Args:
			actionEvent (dragonfly.ActionEvent): After ActionEvent.
		"""
		self.__after.append(actionEvent)

	def __doEvent(self, action: "dragonfly.Action", eventList: typing.List["dragonfly.ActionEvent"]) -> bool:
		"""Perform ActionEvent match with the Action, check if it meets the condition, and executes
		the event.

		eventList argument allows especify Before or After events.

		Args:
			action (dragonfly.Action): target Action.
			eventList (typing.List[action.ActionEvent]): list of the events.

		Returns:
			bool: True if and only if the event has not cacelled.
		"""
		result = True
		for actionEvent in eventList:
			# If match action with list
			if actionEvent.match(action):
				# Check if actionevent's conditions return true
				if not actionEvent.checkConditions(action): continue
				
				# Execute responses
				actionEvent.execute(action)
				result = not actionEvent.cancel

		return result

	def doBefore(self, action: "dragonfly.Action") -> bool:
		"""Perform Before ActionEvent match with the Action, check if it meets the condition, and executes
		the event.

		Args:
			action (action.Action): target Action.

		Returns:
			bool: True if and only if the event has not cacelled.
		"""
		return self.__doEvent(action, self.__before)

	def doAfter(self, action: "dragonfly.Action") -> bool:
		"""Perform After ActionEvent match with the Action, check if it meets the condition, and executes
		the event.

		Args:
			action (action.Action): target Action.

		Returns:
			bool: True if and only if the event has not cacelled.
		"""
		return self.__doEvent(action, self.__after)

	@property
	def the(self) -> str:
		"""Construct the first name of the noun adding the definited article depending the
		genre and the number of self.

		Returns:
			str: The name with the article.
		"""
		result = ""

		# If is a propper noun:
		if self.isSet("propper"):
			result = self.name
		else:
			for a in self.dictionary.articles():
				if a.female == self.isSet("female") and a.plural == self.isSet("plural") and not a.indefinited:
					result = a.name + " " + self.name

		return result

	@property
	def a(self) -> str:
		"""Construct the first name of the noun adding the indefinited article depending the
		genre and the number of self. If the noun is countless, return only the first name.

		Returns:
			str: The name with the article.
		"""

		if self.isSet("countless"): return self.name

		result = ""
		for a in self.dictionary.articles():
			if a.female == self.isSet("female") and a.plural == self.isSet("plural") and a.indefinited:
				result = a.name + " " + self.name

		return result

	@property
	def article(self) -> str:
		"""Return the definited article if the noun is definited, for otherwise return the
		indefinited article.

		Returns:
			str: the article depending the noun.
		"""
		return self.the if self.isSet("definited") else self.a

	@property
	def connections(self) -> typing.List["dragonfly.Connection"]:
		"""Return the list of all connections of the noun.

		Returns:
			typing.List[dragonfly.Connection]: the list of the connections.
		"""
		return self.__connections

	def addConnection(self, conn: "dragonfly.Connection") -> None:
		"""Add a new connection to noun. If the exit exists, replace the destiny with the new instance.

		Args:
			conn (dragonfly.Connection): a new connection.
		"""

		exit = self.dictionary.exit(conn.exit)

		if not exit:
			raise dragonfly.DragonflyException(
				f'Adding connection: Exit "{conn.exit}" not found in dictionary.')

		for c in self.connections:
			if exit.responds(c.exit):
				c.destiny = conn.destiny
				return None

		self.__connections.append(conn)

	def connection(self, exit: "dragonfly.Exit") -> "dragonfly.Connection":
		"""Return a connection indicating the associated exit.

		Args:
			exit (Exit): The exit to match connection.

		Returns:
			dragonfly.Connection: the connection associated to exit. If not exists any connection
							with the exit, return None.
		"""
		for c in self.__connections:
			if exit.responds(c.exit): return c
		return None

	def clone(self, container: "Noun") -> None:
		"""Clone the object on dictionary inner especified container.

		Utilice clone to add generic objects like floor, walls and roof.

		Args:
			container (Noun): Destiny for cloned instance.
		"""
		noun = copy.copy(self)
		noun.names = self.names
		noun.game = self.game
		noun.container = container

		self.dictionary.addNoun(noun)

		noun.set(self.attributes)

		for k in self.variables:
			noun.setVariable(self.getVariable(k))

		for n in self.childs():
			n.clone(noun)

	def load(self, element: QDomElement):
		"""Load the noun from xml element.

		Args:
			element (QDomElement): xml elememt.
		"""
		super().load(element)

		for i in range(element.childNodes().count()):
			node = element.childNodes().at(i)
			if node.nodeType() == QDomNode.ElementNode:
				child = node.toElement()

				# Sets
				if child.nodeName() == "set":
					for j in range(child.childNodes().count()):
						n = child.childNodes().at(j)
						if n.nodeType() == QDomNode.TextNode:
							sets = n.toText().data()
							for s in sets.split(","):
								self.set([s.strip()])

				# Variables
				if child.nodeName() == "variable":
					name = child.attribute("name")
					value = child.attribute("value")
					self.setVariable(name, value)

				# Childs
				if child.nodeName() == "noun":
					childNoun = Noun(container=self)
					childNoun.game = self.game
					childNoun.load(child)
					self.dictionary.addNoun(childNoun)

				# Events
				if child.nodeName() == "before":
					event = dragonfly.ActionEvent()
					event.load(child)
					self.addBefore(event)

				if child.nodeName() == "after":
					event = dragonfly.ActionEvent()
					event.load(child)
					self.addAfter(event)

				# Connections
				if child.nodeName() == "connection":
					conn = dragonfly.Connection()
					conn.load(child)
					self.addConnection(conn)

				# Clones
				if child.nodeName() == "clone":
					instance = child.attribute("instance")
					lst = self.dictionary.nouns(instance)
					if not lst:
						raise dragonfly.DragonflyException(
							f'Noun target "{instance}" not found in dictionary.')
					cl = lst[0]

					cl.clone(self)

File: base/noun.js
import { DFMLElement } from "../dfml/js/main/element.js";
import { DFMLNode } from "../dfml/js/main/node.js";
import { DFMLValue } from "../dfml/js/main/value.js";
import { Action } from "./action.js";
import { Entity } from "./entity.js";
import { ActionEvent } from "./actionevent.js";
import { Output } from "./output.js";
import { Connection } from "./movement.js";
import { Utils } from "./utils.js";

/**
 * Nouns represents the objects of the game. Can be contained by other nouns.
 *		and has attributes and variables.
 *
 * @export
 * @class Noun
 * @extends {Element}
 */
export class Noun extends Entity {
	static #idMax = 0;

	/**
	 * Creates an instance of Noun.
	 * @param {Noun} container
	 * @memberof Noun
	 */
	constructor(container = null) {

		super();

		this.container = null;

		if (container != undefined) this.container = container;

		this.book = null;
		this.dictionary = null;

		Noun.#idMax ++;
		this.id = Noun.#idMax;

		this.attrs = new Set();
		this.variables = {};

		this.beforeEvents = [];
		this.afterEvents = [];

		this.connections = [];
	}

	/**
	 * Check if this noun contains a noun.
	 *
	 * @param {string} name the name of the child.
	 * @return {boolean} true if and only if this contains the child.
	 * @memberof Noun
	 */
	contains(name) {

		for (const n of this.dictionary.nouns) {
			if (n.container === this) {
				if (n.responds(name)) return true;
			}
		}

		return false;
	}

	/**
	 * Return a list with chils wich responds to name.
	 * If the name = "", return all childs.
	 *
	 * @param {string} name
	 * @memberof Noun
	 */
	children(name) {
		let result = [];
		this.dictionary.nouns.forEach(n => {
			if (n.container == this) {
				if (name === undefined) {
					result.push(n);
				} else {
					if (n.responds(name)) result.push(n);
				}
			}
		});
		return result;
	}

	/**
	 * Set a list of attributes.
	 *
	 * @param {List} values
	 * @memberof Noun
	 */
	set(values) {
		this.attrs = new Set([...this.attrs, ...values]);
	}

	/**
	 * Check if an attribute is setted.
	 */
	isSet(value) {
		return this.attrs.has(value);
	}

	/**
	 * Unset a list of attributes.
	 */
	unset(values) {
		values.forEach(v => {
			this.attrs.delete(v);
		});
	}


	/**
	 * Return the value of the variable.
	 *
	 * @param {string} name name of the variable.
	 * @return {string} the value of the variable.
	 * @memberof Noun
	 */
	getVariable(name) {
		if (name in variables) {
			return variables[name];
		}
		
		return "";
	}

	/**
	 * Set the value of the variable.
	 *
	 * @param {string} name name of the variable.
	 * @param {string} value value of the variable.
	 * @memberof Noun
	 */
	setVariable(name, value) {
		this.variables[name] = value;
	}

	/**
	 * Add a new Before action evento to the list.
	 *
	 * @param {ActionEvent} actionEvent Before ActionEvent.
	 * @memberof Noun
	 */
	addBefore(actionEvent) {
		this.beforeEvents.push(actionEvent);
	}

	/**
	 * Add a new After action evento to the list.
	 *
	 * @param {ActionEvent} actionEvent After ActionEvent.
	 * @memberof Noun
	 */
	addAfter(actionEvent) {
		this.afterEvents.push(actionEvent);
	}

	/**
	 * Perform ActionEvent match with the Action, check if it meets the condition, and executes
	 * the event.

	 * eventList argument allows especify Before or After events.
	 *
	 * @param {Action} action target Action.
	 * @param {Array.<ActionEvent>} eventList list of the events.
	 * @return {boolean} 
	 * @memberof Noun
	 */
	async doEvent(action, eventList) {
			
		let result = true;
		for (const actionEvent of eventList) {
			// If match action with list
			if (actionEvent.match(action)) {
				// Check if actionevent's conditions return true
				if (!actionEvent.checkConditions(action)) continue;
				
				// Execute responses
				await actionEvent.execute(action);
				result = !actionEvent.cancel;
			}
		}

		return result;
	}

	/**
	 * Perform Before ActionEvent match with the Action, check if it meets the condition, and executes
     * the event.
	 *
	 * @param {Action} action target Action.
	 * @return {boolean} True if and only if the event has not cacelled.
	 * @memberof Noun
	 */
	async doBefore(action) {
		return await this.doEvent(action, this.beforeEvents);
	}
	
	/**
	 * Perform After ActionEvent match with the Action, check if it meets the condition, and executes
	 * the event.
	 *
	 * @param {Action} action target Action.
	 * @return {boolean}  True if and only if the event has not cacelled.
	 * @memberof Noun
	 */
	async doAfter(action) {
		return await this.doEvent(action, this.afterEvents);
	}

	/**
	 * Construct the first name of the noun adding the definite article depending on the
	 * gender and the number of `this`.
	 *
	 * @return {string} The name with the article.
	 * @memberof Noun
	 */
	the() {
        
        let result = "";

        // If it's a proper noun:
        if (this.isSet("proper")) {
            result = this.getName();
        } else {
            for (const a of this.dictionary.articles) {
                if (a.female === this.isSet("female") && a.plural === this.isSet("plural") && !a.indefinited) {
                    result = `${a.name} ${this.getName()}`;
                    break;
                }
            }
        }

        return result;
    }

	/**
	 * Construct the first name of the noun adding the indefinite article depending on the
	 * gender and the number of `this`. If the noun is countless, return only the first name.
	 *
	 * @return {string} The name with the article.
	 * @memberof Noun
	 */
	a() {
        if (this.isSet("countless")) return this.getName();

        let result = "";
        for (const a of this.dictionary.articles) {
            if (a.female === this.isSet("female") && a.plural === this.isSet("plural") && a.indefinited) {
                
				if (a.name === 'a' && this.isSet('an')) {
					result = `an ${this.getName()}`;
					break;
				}
				
				result = `${a.name} ${this.getName()}`;
                break;
            }
        }

        return result;
    }

	/**
	 * Return the definite article if the noun is defined; otherwise, return the indefinite article.
	 *
	 * @return {string} The article depending on the noun.
	 * @memberof Noun
	 */
	article() {
        return this.isSet("definited") ? this.the() : this.a();
    }

	/**
	 * Return a connection indicating the associated exit.
	 *
	 * @param {string} exit The exit to match connection.
	 * @return {Connection} the connection associated to exit. If not exists any connection
	 *					with the exit, return null.
	 * @memberof Noun
	 */
	getConnection(exit) {
		let conn = null;

		const exitObj = this.dictionary.getExit(exit);
		if (!exitObj) {
			Output.error(`Get connection from "${this.getName()}": exit "${exit}" not found in dictionary.`);
			return null;
		}

		this.connections.forEach((c) => {
			if (exitObj.responds(c.exit)) conn = c;
		});

		return conn;
	}

	/**
	 * Remove connection giving the exit.
	 *
	 * @param {string} exit associated exit to connection to remove.
	 * @memberof Noun
	 */
	removeConnection(exit) {

		const exitObj = this.dictionary.getExit(exit);
		if (!exitObj) {
			Output.error(`Removing connection from "${this.getName()}": exit "${exit} not found in dictionary."`);
			return ;
		}

		// Gets the index
		let index = 0;
		for (const conn of this.connections) {
			if (exitObj.responds(conn.exit)) break;
			index ++;
		}

		// Connection not found
		if (index >= this.connections.length) {
			Output.error(`Removing connection from "${this.getName()}": connection "${exit}" not found.`)
			return ;
		}

		// Remove connection
		this.connections.splice(index, 1);

	}

	/**
	 * Clone the object on dictionary inner especified container.
	 * Utilice clone to add generic objects like floor, walls and roof.
	 *
	 * @param {Noun} container
	 * @memberof Noun
	 */
	clone(container) {
		const copy = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
		copy.container = container;

		this.dictionary.nouns.push(copy);

	}

	/**
	 * Load the noun from dfml element.
	 *
	 * @param {DFMLNode} node dfml element.
	 * @memberof Noun
	 */
	load(node) {
		super.load(node);

		node.getChildren().forEach(e => {
			if (e.getElementType() === DFMLElement.NODE) {
				if (e.getName() === "set") {
					const children = e.getChildren();
					children.forEach(c => {
						if (c.getElementType() === DFMLElement.DATA &&
							c.getValue().getType() === DFMLValue.STRING) {
							this.set([ c.getValue().getValue() ]);
						} else {
							Output.error(`(Set) on load noun '${this.getName()}': no string data attribute.`);
						}
					});
				} else if (e.getName() === "variable") {

					if (Utils.expectedAttributes(e, "name", "value")) {
						this.setVariable(e.getAttr("name").getValue(),
						e.getAttr("value").getValue());
					}
				} else if (e.getName() === "noun") {
					const noun = new Noun(this);
					noun.book = this.book;
					noun.dictionary = this.dictionary;
					noun.load(e);
					this.dictionary.nouns.push(noun);
				} else if (e.getName() === "before") {
					const event = new ActionEvent();
					event.load(e);
					this.beforeEvents.push(event);
				} else if (e.getName() === "after") {
					const event = new ActionEvent();
					event.load(e);
					this.afterEvents.push(event);
				} else if (e.getName() === "connection") {
					const connection = new Connection();
					connection.load(e);
					this.connections.push(connection);
				} else if (e.getName() === "describe-place") {
					e.setName("after");
					e.setAttrString("actions", "LookAround");
					const event = new ActionEvent();
					event.load(e);
					this.afterEvents.push(event);
				} else if (e.getName() === "describe-object") {
					e.setName("after");
					e.setAttrString("actions", "ExamineObject");
					const event = new ActionEvent();
					event.load(e);
					this.afterEvents.push(event);
				} else if (e.getName() === "clone") {

					if (Utils.expectedAttributes(e, "instance")) {

						const instance = e.getAttr("instance").getValue();
						const lst = this.dictionary.getNouns(instance);

						if (lst.length === 0) {
							Output.error(`Clone: noun target "${instance}" not found in dictionary.`);
							return ;
						}

						lst[0].clone(this);

					}

				}
			}
		});
	}
}
